include the command being run in the default title

adqm opened the pull request:
This is a little modification (again, another idea borrowed from fish) to make the console title, by default, show the name of the command being run (the current active foreground job).  I often have a bunch of terminals open, and this makes it easier for me to switch between them.

Normal title on my machine is:
`hartz@compy-488: ~ | xonsh`

Opening, e.g., `python` makes it:
`python | hartz@compy-488: ~ | xonsh`

Programs like `vim` that set titles themselves are unaffected.

Also works with pipelines.  So opening `echo "hello" | gedit -` makes it:
`gedit | hartz@compy-488: ~ | xonsh`

Tested to work with `fg`, `bg`, and friends.

scopatz commented on xonsh/base_shell.py:
```diff
@@ -203,6 +203,7 @@ def settitle(self):
             os.system('title {}'.format(t))
         else:
             sys.stdout.write("\x1b]2;{0}\x07".format(t))
+            sys.stdout.flush()
```
surprised that wasn't there already.

scopatz commented on xonsh/built_ins.py:
```diff
@@ -613,6 +613,10 @@ def run_subproc(cmds, captured=True):
+    try:
+        builtins.__xonsh_shell__.settitle()
+    except AttributeError:
+        pass
```
This block could be:

```python
if hasattr(builtins.__xonsh_shell__, 'settitle'):
    builtins.__xonsh_shell__.settitle()
```

rather than relying on exceptions.  Or, just remove any conditional and do a straight up `builtins.__xonsh_shell__.settitle()` and rely on duck typing.

adqm replied:
Without the conditional (or the try/except), the `import` tests failed, so I stuck this in.

The problem is actually `builtins` not having `__xonsh_shell__` in that case, not `__xonsh_shell__` not having `settitle`.

I also tried setting a default (`None`) for `__xonsh_shell__` in `builtins.py`, but that didn't work, either (same error).  So maybe imports don't even initialize the builtins?  Not sure what was happening.  Or maybe it's more a quirk of `nosetests3` than anything else?

scopatz replied:
Ahh ok.  Yeah, that makes more sense.  Some of the tests mock the builtins initialization. I think that code is in `tests/tools.py`. As I recall, the `__xonsh_shell__` came later. So it might not be mocked.

scopatz commented on xonsh/environ.py:
```diff
@@ -580,6 +589,7 @@ def _collapsed_pwd():
     short_cwd=_collapsed_pwd,
     curr_branch=current_branch,
     branch_color=branch_color,
+    current_job=_current_job,
```
Can you add this to the relevant `docs/tutorial.rst` section so that people know about it.

adqm replied:
Can do.

scopatz commented:
Looks great @adqm! Thanks a ton.  Just a few small issues here and there.  The big one is documentation.

adqm commented:
Alright, I think it's ready for another look.

I added a dummy shell to the mock environment in `tests/tools.py` so that the tests won't fail anymore, and I replaced the try/except with a check of `$XONSH_INTERACTIVE` (since it doesn't make much sense to set the title if we're not running interactively).

I also added a very short entry for `current_job` to the `$PROMPT` section of the tutorial.

scopatz commented:
This doesn't seem to work super well `$XONSH_STORE_STDOUT=True`.  The name of the program with flicker in the title at the *end* of the command.  Even something like `sleep 2` can demonstrate this.  Maybe there is another place to in `run_subproc()` to put the call to `settitle()`, or maybe we should be writing to the `prev_proc.stdout` in this case.

I am happy to merge it in as-is, because it does work when this variable is False, if you want to open up an another issue.

adqm commented:
Hmmm, that's interesting.  Let's hold off on merging, and I'll take another look.

scopatz commented:
Sounds good.  I am pretty sure that this is because TeePTY forks the process and thus changes stdin, stdout, and stderr.

adqm commented:
Okay, so `TeePTY` looks like it actually holds on to the main thread, rather than running in the background like `Popen` or `ProcProxy` (hence the wait time and title only flashing briefly).  I'm going to try to "fix up" `TeePTY` so that it behaves a bit more like those other two.  Failing that, we can maybe just merge in as it currently stands (which doesn't try to do that second `settitle` if we're capturing stdout).

scopatz commented:
> I'm going to try to "fix up" TeePTY so that it behaves a bit more like those other two. 

Because of the process fork, I think that TeePTY needs to run on the main thread.  That said there may be another way to fix it up.  Such a fix would probably also address #519.

scopatz commented:
Or might address #519.

adqm commented:
Okay, so since there is at least one other thing that is broken with `TeePTY`, my vote is to merge this in now (if it gets your seal of approval), and to tackle `TeePTY` in another PR (hopefully addressing the remaining issue here, as well as #519).

scopatz commented:
OK. Sounds good.

scopatz commented:
Thanks @adqm!
